lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs

```diff
@@ -196,6 +222,43 @@ newStakePoolLayer nl db@DBLayer {..} =
     _knownPools =
         Set.fromList <$> liftIO (atomically listRegisteredPools)
 
+    -- In order to apply the settings, we have to restart the
+    -- metadata sync thread as well to avoid race conditions.
+    _putSettings :: MVar (Async ()) -> MVar Bool -> Settings -> IO ()
+    _putSettings mas mv settings = do
+        stopSyncThread
+        gcMetadata
+        writeSettings
+        restartSyncThread
+      where
+        -- stop syncing thread, so we can apply the settings cleanly
+        stopSyncThread = do
+            void $ swapMVar mv True
+            as <- readMVar mas
+            void $ waitCatch as
+
+        -- clean up metadata table if the new sync settings suggests so
+        gcMetadata = do
+            oldSettings <- liftIO $ atomically $ readSettings
+            case (poolMetadataSource oldSettings, poolMetadataSource settings) of
+                (_, FetchNone) -> -- this is necessary if it's e.g. the first time
+                                  -- we start the server with the new feature
+                                  -- and the database has still stuff in it
```

Comment:
I still do not understand why this isn't already captured by `old /= new`. Surely, if the database already has stuff in it, it also has a given setting that is not `FetchNone`?

Reply:
This is the case of e.g. upgrading the wallet from a previous version. The default setting is `FetchNone`, but the database is already populated. Tbf... we might wanna remove this clause and somehow make sure that the database migration wipes out the table? There's more cases that can go wrong. This was the most obvious one that broke local testing.